images/image.go

```diff
@@ -406,3 +444,420 @@ func IsCompressedDiff(ctx context.Context, mediaType string) (bool, error) {
 	}
 	return false, nil
 }
+
+// encryptLayer encrypts the layer using the CryptoConfig and creates a new OCI Descriptor.
+// A call to this function may also only manipulate the wrapped keys list.
+// The caller is expected to store the returned encrypted data and OCI Descriptor
+func encryptLayer(cc *encconfig.CryptoConfig, dataReader content.ReaderAt, desc ocispec.Descriptor) (ocispec.Descriptor, io.Reader, error) {
+	var (
+		size int64
+		d    digest.Digest
+		err  error
+	)
+
+	encLayerReader, annotations, err := encryption.EncryptLayer(cc.EncryptConfig, encryption.ReaderFromReaderAt(dataReader), desc)
+	if err != nil {
+		return ocispec.Descriptor{}, nil, err
+	}
+
+	// were data touched ?
+	if encLayerReader != nil {
+		size = 0
+		d = ""
+	} else {
+		size = desc.Size
+		d = desc.Digest
+	}
+
+	newDesc := ocispec.Descriptor{
+		Digest:      d,
+		Size:        size,
+		Platform:    desc.Platform,
+		Annotations: annotations,
+	}
+
+	switch desc.MediaType {
+	case MediaTypeDockerSchema2LayerGzip:
+		newDesc.MediaType = MediaTypeDockerSchema2LayerGzipEnc
+	case MediaTypeDockerSchema2Layer:
+		newDesc.MediaType = MediaTypeDockerSchema2LayerEnc
+	case MediaTypeDockerSchema2LayerGzipEnc:
+		newDesc.MediaType = MediaTypeDockerSchema2LayerGzipEnc
+	case MediaTypeDockerSchema2LayerEnc:
+		newDesc.MediaType = MediaTypeDockerSchema2LayerEnc
+
+		// TODO: Mediatypes to be added in ocispec
+	case ocispec.MediaTypeImageLayerGzip:
+		newDesc.MediaType = MediaTypeDockerSchema2LayerGzipEnc
+	case ocispec.MediaTypeImageLayer:
+		newDesc.MediaType = MediaTypeDockerSchema2LayerEnc
+
+	default:
+		return ocispec.Descriptor{}, nil, errors.Errorf("Encryption: unsupporter layer MediaType: %s\n", desc.MediaType)
+	}
+
+	return newDesc, encLayerReader, nil
+}
+
+// DecryptBlob decrypts the layer blob using the CryptoConfig and creates a new OCI Descriptor.
+// The caller is expected to store the returned plain data and OCI Descriptor
+func DecryptBlob(cc *encconfig.CryptoConfig, dataReader content.ReaderAt, desc ocispec.Descriptor, unwrapOnly bool) (ocispec.Descriptor, io.Reader, error) {
+	if cc == nil {
+		return ocispec.Descriptor{}, nil, errors.Wrapf(errdefs.ErrInvalidArgument, "CryptoConfig must not be nil")
+	}
+	resultReader, err := encryption.DecryptLayer(cc.DecryptConfig, encryption.ReaderFromReaderAt(dataReader), desc, unwrapOnly)
+	if err != nil || unwrapOnly {
+		return ocispec.Descriptor{}, nil, err
+	}
+
+	newDesc := ocispec.Descriptor{
+		Size:     0,
+		Platform: desc.Platform,
+	}
+
+	switch desc.MediaType {
+	case MediaTypeDockerSchema2LayerGzipEnc:
+		newDesc.MediaType = MediaTypeDockerSchema2LayerGzip
+	case MediaTypeDockerSchema2LayerEnc:
+		newDesc.MediaType = MediaTypeDockerSchema2Layer
+	default:
+		return ocispec.Descriptor{}, nil, errors.Errorf("Decryption: unsupporter layer MediaType: %s\n", desc.MediaType)
+	}
+	return newDesc, resultReader, nil
```

Comment:
Seems weird to return a descriptor from a public function without a digest. Is the media type the only value needed?